### ISSUE TITLE
Validation training isn't supported by the new API

### ISSUE THREAD

**Issue by rogancarr** [dotnet/machinelearning#2502]
I cannot find a way to train with a validation set in the new API, and I think there is some confusion as to how this would work, were it possible.

Take this example:

```cs
// Create a pipeline to train on the sentiment data
var trainData = readTrainData();
var validData = readValidData();

var pipeline = mlContext.Transforms.SomTransform.()
    .AppendCacheCheckpoint(mlContext) as IEstimator<ITransformer>;
var preprocessor = pipeline.Fit(trainData);
var preprocessedValidData = preprocessor.Transform(validData);

// Train model with validation set.
// There is no way below to specify a validation set for the learner
pipeline = pipeline.Append(mlContext.Regression.Trainers.FastTree(numTrees: 2));
// Nor is there a way to specify a validation set in the Fit
var model = pipeline.Fit(trainData);
```

This example uses `FastTree`, but the same problem exists for `GAM` and `FFM`, our other validation-set trainers: There is nowhere to specify a validation set.

A second problem exists, namely: If we were to specify a validaiton set, would we specify the raw, unprocessed validation set (here `validData`) or would we specify the pre-transformed validation set (here, `preprocessedValidData`).

**Comment by Ivanidzo4ka** [comment]
`mlContext.Regression.Trainers.FastTree(numTrees: 2).Train(trainData, preprocessedValidData)`

**Comment by TomFinley** [comment]
Why `Train`, not `Fit`? This would be more consistent, make them easier to find, etc.?

**Comment by Ivanidzo4ka** [comment]
Sure. All I wanted to say, it's possible, it just can't be part of IEstimatorChain, it should be separate call on trainer itself.

**Comment by rogancarr** [comment]
@Ivanidzo4ka @TomFinley Ah, I thought I was missing something &mdash; thanks for pointing it out.

New question: How do I save my pipeline out for scoring now that it's in different pieces?
- The featurization pipeline
- The training pipeline

And what do we do if this is just one part of a larger pipeline?

Do I then have to stitch it together by hand before I can save it? (If this is possible?)

**Comment by rogancarr** [comment]
To answer my own question, we can do this programmatically: 

```cs
// Preprocess the datasets
var preprocessor = pipeline.Fit(trainData);
var preprocessedTrainData = preprocessor.Transform(trainData);
var preprocessedValidData = preprocessor.Transform(validData);

// Train the model with a validation set
var trainedModel = mlContext.Regression.Trainers.FastTree(numTrees: 2)
    .Train(trainData: preprocessedTrainData, validationData: preprocessedValidData);

// Combine the model
var model = preprocessor.Append(trainedModel);
```

I think this is probably okay for a V1.0 release, but it would be nicer to have a formulation of a training pipeline as a DAG, where we can have multiple inputs (e.g. one or more training sets, validation sets, etc.).

**Comment by TomFinley** [comment]
> I think this is probably okay for a V1.0 release, but it would be nicer to have a formulation of a training pipeline as a DAG, where we can have multiple inputs (e.g. one or more training sets, validation sets, etc.).

We already have a DAG. You have variables with values you can assign, pass those variables into methods, and so forth. The good news is, it's actually a lot more powerful than a DAG... you cans have loops, write your own classes to encapsulate functionality... 😄 

Seriously, the fact that you answered your own question with the conclusion being that you had to remember you were operating in C# and could therefore assign variables, call methods, and so on, is absolutely the right answer to this question. Frankly, I think it would be kind of silly for us to re-invent such basic concepts for this API. That usage of our C# .NET API for machine learning winds up looking like C# code is good, not bad.
